sonic-xcvrd/xcvrd/sff_mgr.py

```diff
@@ -0,0 +1,449 @@
+        self.port_dict = {}
+        # port_dict snapshot captured in the previous event update loop
+        self.port_dict_prev = {}
+        self.xcvr_table_helper = XcvrTableHelper(namespaces)
+        self.namespaces = namespaces
+
+    def log_notice(self, message):
+        self.helper_logger.log_notice("SFF: {}".format(message))
+
+    def log_warning(self, message):
+        self.helper_logger.log_warning("SFF: {}".format(message))
+
+    def log_error(self, message):
+        self.helper_logger.log_error("SFF: {}".format(message))
+
+    def get_active_lanes_for_lport(self, lport, subport_idx, num_lanes_per_lport, num_lanes_per_pport):
+        """
+        Get the active lanes for a logical port based on the subport index.
+
+        Args:
+            lport (str): Logical port name.
+            subport_idx (int): Subport index, starting from 1. 0 means all lanes are taken.
+            num_lanes_per_lport (int): Number of lanes per logical port.
+            num_lanes_per_pport (int): Number of lanes per physical port.
+
+        Returns:
+            list: A list of boolean values, where True means the corresponding
+                  lane is active.
+        """
+        if subport_idx < 0 or subport_idx > num_lanes_per_pport // num_lanes_per_lport:
+            self.log_error(
+                f"{lport}: Invalid subport_idx {subport_idx} "
+                f"for num_lanes_per_lport={num_lanes_per_lport}, "
+                f"num_lanes_per_pport={num_lanes_per_pport}"
+            )
+            return None
+
+        if subport_idx == 0:
+            lanes = [True] * num_lanes_per_pport
+        else:
+            lanes = [False] * num_lanes_per_pport
+            start = (subport_idx - 1) * num_lanes_per_lport
+            end = subport_idx * num_lanes_per_lport
+            lanes[start:end] = [True] * (end - start)
+
+        return lanes
+
+    def on_port_update_event(self, port_change_event):
+        if (port_change_event.event_type
+                not in [port_change_event.PORT_SET, port_change_event.PORT_DEL]):
+            return
+
+        lport = port_change_event.port_name
+        pport = port_change_event.port_index
+        asic_id = port_change_event.asic_id
+
+        # Skip if it's not a physical port
+        if not lport.startswith('Ethernet'):
+            return
+
+        # Skip if the physical index is not available
+        if pport is None:
+            return
+
+        if port_change_event.port_dict is None:
+            return
+
+        if port_change_event.event_type == port_change_event.PORT_SET:
+            if lport not in self.port_dict:
+                self.port_dict[lport] = {}
+            if pport >= 0:
+                self.port_dict[lport]['index'] = pport
+
+            if self.SUBPORT in port_change_event.port_dict and \
+                self.LANES_LIST in port_change_event.port_dict:
+                subport_idx = int(port_change_event.port_dict[self.SUBPORT])
```

Comment:
@longhuan-cisco where is this defined: subport?

Reply:
Hi, is `subport` only used for dynamic port breakout?  I don't see dynamic port breakout handle this field, so, I am not sure that the field is fully supported by sonic.  Or is `subport` used for locating the lanes mapped to logical port? In that case, I think field `lanes` can do that.  So, I am not fully understand why we need subport here. Could you please kindly elaborate?

Reply:
'subport' is used for breakout situation,`subport` field determines which lanes belongs to which subport, based on which tx_disable/enable API (which takes lane mask as argument) can be called per subport.
The `lanes` field in CONFIG_DB are just serdes lane numbers (e.g. 2304). Those are just some big HW level number, and it can't derive the lane mask.

today's cmis_mgr is already relying on this `subport`. today's DPB HLD has a section for `subport` also.
For now, it's supported only for static breakout config via port_config.ini.

example:
Assuming, we have a physical port with serdes `lanes` (2304,2305,2306,2307), then we do a 4x breakout as below,
  - `subport` 1 (serdes `lanes`=2304): 1st lane   (lane mask 0b0001 for tx_disable/enable API)
  - `subport` 2 (serdes `lanes`=2305): 2nd lane (lane mask 0b0010)
  - `subport` 3 (serdes `lanes`=2306): 3rd lane  (lane mask 0b0100)
  - `subport` 4 (serdes `lanes`=2307): 4th lane.   (lane mask 0b1000)

Reply:
Thanks for the explanation. So, dynamic port breakout is not fully supported, but static port breakout is supported by this field. @keboliu , @eddyk-nvidia for awareness.